### ISSUE TITLE
DataFrame.reparition raises not user-friendly exception on non-indexed DataFrame

### ISSUE THREAD

**Issue by frol** [dask/dask#1380]
At the very least there should be a useful error. At the best, I would love to use `repartition` without having to set an index.

```python
>>> df = dask.dataframe.read_csv('qq.csv')
>>> df.repartition(npartitions=1)

/miniconda3/envs/dask/lib/python3.5/site-packages/dask/dataframe/core.py in repartition(self, divisions, npartitions, force)
    489             if npartitions > self.npartitions:
    490                 raise ValueError("Can only repartition to fewer partitions")
--> 491             return repartition_npartitions(self, npartitions)
    492         elif divisions is not None:
    493             return repartition(self, divisions, force=force)

/miniconda3/envs/dask/lib/python3.5/site-packages/dask/dataframe/core.py in repartition_npartitions(df, npartitions)
   2636     if len(divisions) <= npartitions:
   2637         divisions = divisions + (df.divisions[-1],)
-> 2638     return df.repartition(divisions=divisions)
   2639 
   2640 

/miniconda3/envs/dask/lib/python3.5/site-packages/dask/dataframe/core.py in repartition(self, divisions, npartitions, force)
    491             return repartition_npartitions(self, npartitions)
    492         elif divisions is not None:
--> 493             return repartition(self, divisions, force=force)
    494         else:
    495             raise ValueError(

/miniconda3/envs/dask/lib/python3.5/site-packages/dask/dataframe/core.py in repartition(df, divisions, force)
   2673         out = 'repartition-merge-' + token
   2674         dsk = repartition_divisions(df.divisions, divisions,
-> 2675                                     df._name, tmp, out, force=force)
   2676         return df._constructor(merge(df.dask, dsk), out,
   2677                                df._pd, divisions)

/miniconda3/envs/dask/lib/python3.5/site-packages/dask/dataframe/core.py in repartition_divisions(a, b, name, out1, out2, force)
   2522         raise ValueError('New division must be longer than 2 elements')
   2523 
-> 2524     if b != sorted(b):
   2525         raise ValueError('New division must be sorted')
   2526     if len(b[:-1]) != len(list(unique(b[:-1]))):

TypeError: unorderable types: NoneType() < NoneType()
```

**Comment by jcrist** [comment]
My proposed behavior is:

- If given `npartitions`, should work with both indexed and non-indexed dataframes
- If given `divisisions`, should raise an informative error if not already indexed. I like this rather than potentially doing an expensive operation without the user explicitly asking for it. Alternatively, a keyword (`set_index=True`, default is False, potentially with a better keyword name) might be used to note that repartition can index if needed.

Thoughts?

**Comment by mrocklin** [comment]
I'm in favor of pointing people towards using `set_index` explicitly rather than doing it for them.

**Comment by mrocklin** [comment]
Other than that I'm +1 on @jcrist 's suggestions

**Comment by mrocklin** [comment]
Fixed by #1396
